### PR TITLE
Add reward and context inputs

### PR DESCRIPTION
**Pull request by Thoralf-M**
# Description of change

Add reward context input https://github.com/iotaledger/tips-draft/blob/tip45/tips/TIP-0045/tip-0045.md#context-inputs

## Links to any relevant issues

Closes https://github.com/iotaledger/iota-sdk/issues/909
Closes https://github.com/iotaledger/iota-sdk/issues/910

## Type of change

- Enhancement (a non-breaking change which adds functionality)

### CODE REVIEW

File: sdk/src/types/block/context_input/reward.rs [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+use core::str::FromStr;
+
+use derive_more::From;
+
+use crate::types::block::Error;
+
+/// A Reward Input is an input that indicates which transaction Input is claiming Mana rewards.
+#[derive(Clone, Copy, Eq, PartialEq, Hash, Ord, PartialOrd, From, packable::Packable)]
+pub struct RewardInput(u16);
```
Comment (thibault-martinez) [inline sdk/src/types/block/context_input/reward.rs]: For consistency with our other enum variants, this should be called `RewardContextInput`

Comment (Thoralf-M) [reply sdk/src/types/block/context_input/reward.rs]: Done

File: sdk/src/types/block/context_input/mod.rs [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+impl ContextInput {
+    /// Returns the input kind of an `ContextInput`.
+    pub fn kind(&self) -> u8 {
+        match self {
+            Self::Reward(_) => RewardContextInput::KIND,
+        }
+    }
+
+    /// Checks whether the input is a [`RewardContextInput`].
+    pub fn is_utxo(&self) -> bool {
```
Comment (thibault-martinez) [inline sdk/src/types/block/context_input/mod.rs]: :trollface:

File: sdk/src/types/block/context_input/mod.rs [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+// SPDX-License-Identifier: Apache-2.0
+
+mod reward;
+
+use derive_more::From;
+
+use self::reward::RewardContextInput;
+use crate::types::block::Error;
+
+/// A generic input supporting different input kinds.
```
Comment (thibault-martinez) [inline sdk/src/types/block/context_input/mod.rs]: Do we have a better explanation in the TIP?

File: sdk/src/types/block/context_input/mod.rs (hunk 2 above)
Comment (Thoralf-M) [reply sdk/src/types/block/context_input/mod.rs]: Thanks :trollface:

File: sdk/src/types/block/context_input/mod.rs (hunk 3 above)
Comment (Thoralf-M) [reply sdk/src/types/block/context_input/mod.rs]: Yes, added

File: sdk/src/types/block/context_input/reward.rs [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    }
+
+    /// Returns the index of a [`RewardInput`].
+    pub fn index(&self) -> u16 {
+        self.0
+    }
+}
+
+#[cfg(feature = "serde")]
+string_serde_impl!(RewardContextInput);
```
Comment (thibault-martinez) [inline sdk/src/types/block/context_input/reward.rs]: string impl for an u16?

Comment (Thoralf-M) [reply sdk/src/types/block/context_input/reward.rs]: Also wasn't sure about that, don't we want this for all types?

File: sdk/src/types/block/context_input/mod.rs [hunk 5]
```diff
@@ -0,0 +1,83 @@
+// Copyright 2020-2023 IOTA Stiftung
```
Comment (thibault-martinez) [inline sdk/src/types/block/context_input/mod.rs]:
```suggestion
// Copyright 2023 IOTA Stiftung
```

File: sdk/src/types/block/context_input/reward.rs (hunk 4 above)
Comment (thibault-martinez) [reply sdk/src/types/block/context_input/reward.rs]: Only if they should serialise using their `Display` implementation

File: sdk/src/types/block/context_input/reward.rs [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+use derive_more::From;
+
+use crate::types::block::Error;
+
+/// A Reward Input is an input that indicates which transaction Input is claiming Mana rewards.
+#[derive(Clone, Copy, Eq, PartialEq, Hash, Ord, PartialOrd, From, packable::Packable)]
+pub struct RewardContextInput(u16);
+
+impl RewardContextInput {
+    /// The context input kind of a [`RewardInput`].
```
Comment (thibault-martinez) [inline sdk/src/types/block/context_input/reward.rs]: Rename in the comments as well

File: sdk/src/types/block/context_input/mod.rs [hunk 7]
```diff
@@ -0,0 +1,83 @@
+// Copyright 2023 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+mod reward;
+
+use derive_more::From;
+
+use self::reward::RewardContextInput;
```
Comment (thibault-martinez) [inline sdk/src/types/block/context_input/mod.rs]:
```suggestion
pub use self::reward::RewardContextInput;
```

File: sdk/src/types/block/context_input/reward.rs [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+    impl From<&RewardContextInput> for RewardContextInputDto {
+        fn from(value: &RewardContextInput) -> Self {
+            Self {
+                kind: RewardContextInput::KIND,
+                index: value.index(),
+            }
+        }
+    }
+
+    impl TryFrom<RewardContextInputDto> for RewardContextInput {
```
Comment (thibault-martinez) [inline sdk/src/types/block/context_input/reward.rs]: Looks like this can be a From instead

File: sdk/src/types/block/context_input/mod.rs [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+use crate::types::block::Error;
+
+/// A Context Input provides additional contextual information for the execution of a transaction, such as for different
+/// functionality related to accounts, commitments, or Mana rewards. A Context Input does not need to be unlocked.
+#[derive(Clone, Eq, PartialEq, Hash, Ord, PartialOrd, From, packable::Packable)]
+#[packable(unpack_error = Error)]
+#[packable(tag_type = u8, with_error = Error::InvalidInputKind)]
```
Comment (thibault-martinez) [inline sdk/src/types/block/context_input/mod.rs]: I know they are similar, but maybe it should have its own error for better debugging

File: sdk/src/types/block/context_input/reward.rs [hunk 10]
```diff
@@ -0,0 +1,67 @@
+// Copyright 2023 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use derive_more::From;
+
+/// A Reward Input is an input that indicates which transaction Input is claiming Mana rewards.
```
Comment (qrayven) [inline sdk/src/types/block/context_input/reward.rs]: NIT: `Reward Context Input" ;)

Comment (Thoralf-M) [reply sdk/src/types/block/context_input/reward.rs]: Updated

File: sdk/src/types/block/context_input/mod.rs (hunk 9 above)
Comment (Thoralf-M) [reply sdk/src/types/block/context_input/mod.rs]: Updated

File: sdk/src/types/block/context_input/reward.rs (hunk 8 above)
Comment (Thoralf-M) [reply sdk/src/types/block/context_input/reward.rs]: Updated

File: sdk/src/types/block/context_input/mod.rs [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+impl core::fmt::Debug for ContextInput {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        match self {
+            Self::Reward(input) => input.fmt(f),
+        }
+    }
+}
+
+impl ContextInput {
+    /// Returns the context input kind of an `ContextInput`.
```
Comment (thibault-martinez) [inline sdk/src/types/block/context_input/mod.rs]:
```suggestion
    /// Returns the context input kind of a `ContextInput`.
```

**Review by thibault-martinez — DISMISSED** [review]

File: sdk/src/types/block/context_input/mod.rs [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+    /// Checks whether the context input is a [`RewardContextInput`].
+    pub fn is_reward(&self) -> bool {
+        matches!(self, Self::Reward(_))
+    }
+
+    /// Gets the input as an actual [`RewardContextInput`].
+    /// PANIC: do not call on a non-reward context input.
+    pub fn as_reward(&self) -> &RewardContextInput {
```
Comment (abdulmth) [inline sdk/src/types/block/context_input/mod.rs:46]: can't we add `try_into_reward` and return a `Result` instead?

Comment (Thoralf-M) [reply sdk/src/types/block/context_input/mod.rs:46]: We could, but there are places where we are sure about the type and don't want to handle the Result
We're doing it like this in all the other types already

File: sdk/src/types/block/context_input/reward.rs [hunk 13]
```diff
@@ -0,0 +1,64 @@
+// Copyright 2023 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use derive_more::From;
+
+/// A Reward Context Input is an input that indicates which transaction Input is claiming Mana rewards.
```
Comment (DaughterOfMars) [inline sdk/src/types/block/context_input/reward.rs]:
```suggestion
/// A Reward Context Input indicates which transaction Input is claiming Mana rewards.
```

**Review by DaughterOfMars — DISMISSED** [review]

**Review by abdulmth — DISMISSED** [review]

**Review by abdulmth — DISMISSED** [review]

**Review by thibault-martinez — DISMISSED** [review]

**Review by DaughterOfMars — DISMISSED** [review]

**Review by qrayven — DISMISSED** [review]

File: sdk/src/types/block/context_input/reward.rs [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+}
+
+impl core::fmt::Display for RewardContextInput {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+impl core::fmt::Debug for RewardContextInput {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
```
Comment (kwek20) [inline sdk/src/types/block/context_input/reward.rs]: derive?

**Review by kwek20 — COMMENTED** [review]
Im not sure what our plan is with dtos, if we keep adding 'em, and changing later its a ton of extra effort

File: sdk/src/types/block/context_input/reward.rs (hunk 14 above)
Comment (DaughterOfMars) [reply sdk/src/types/block/context_input/reward.rs]: Actually both this and Display can be derived probably

**Review by DaughterOfMars — APPROVED** [review]

**Review by kwek20 — APPROVED** [review]

**Review by abdulmth — APPROVED** [review]

**Review by thibault-martinez — APPROVED** [review]

**Pull request merged**